tensorflow/core/kernels/ragged_fill_empty_rows_op_gpu.cu.cc

```diff
@@ -0,0 +1,598 @@
+         default_value, nrows, num_empty_rows_host,
+         rows_are_not_ordered_host, first_invalid_index_host,
+         num_empty_rows_through_t, num_empty_rows_through, input_row_ends_t,
+         input_row_ends, empty_row_indicator_t, empty_row_indicator,
+         done]() -> void {
+      DCHECK(done);  // Crash OK
+
+      // Ensure that within the callback, the proper GPU settings are
+      // configured.
+      auto stream = context->op_device_context()->stream();
+      ScopedActivateExecutorContext scoped_activation{stream->parent()};
+
+      int first_invalid_index = *first_invalid_index_host.data();
+      OP_REQUIRES_ASYNC(context, first_invalid_index == kAllValueRowidsValid,
+                        errors::InvalidArgument("valule_rowids(", first_invalid_index,
+                                                ", 0) is invalid."),
+                        done);
+
+      Tindex num_empty_rows = *num_empty_rows_host.data();
+
+      Tindex* output_value_rowids;
+      T* output_values;
+      Tindex* reverse_index_map;
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          AllocateOutputsExceptEmptyRowIndicator(
+              context, N, num_empty_rows, &output_value_rowids, &output_values,
+              &reverse_index_map),
+          done);
+
+      const GPUDevice& device = context->eigen_device<GPUDevice>();
+
+      Tindex* input_index_map = nullptr;
+      Tensor input_index_map_t;
+      int rows_are_not_ordered = *rows_are_not_ordered_host.data();
+      if (rows_are_not_ordered) {
+        OP_REQUIRES_OK_ASYNC(context,
+                             ArgSortByRows(context, device, N, nrows,
+                                           value_rowids, &input_index_map_t),
+                             done);
+        input_index_map = input_index_map_t.vec<Tindex>().data();
+      }
+
+      if (N > 0) {
+        OP_REQUIRES_OK_ASYNC(
+            context,
+            wrap_kernel_call(ScatterInputElementsKernel<T, Tindex>,
+                             /*device=*/device, /*size=*/N, nrows,
+                             input_index_map, value_rowids, values,
+                             num_empty_rows_through, output_value_rowids,
+                             output_values, reverse_index_map),
+            done);
+      }
+
+      if (nrows > 0) {
+        OP_REQUIRES_OK_ASYNC(
+            context,
+            wrap_kernel_call(ScatterNewElementsKernel<T, Tindex>,
+                             /*device=*/device, /*size=*/nrows,
+                             default_value, num_empty_rows_through,
+                             input_row_ends, empty_row_indicator,
+                             output_value_rowids, output_values),
+            done);
+      }
+
+      done();
+    };
+
+    context->device()
+        ->tensorflow_accelerator_device_info()
+        ->event_mgr->ThenExecute(stream, async_finish_computation);
+    return OkStatus();
```

Comment:
Is it possible to share these lines with sparse_fill_empty_rows_op_gpu.cu.cc by passing rank = 1? We could move the common functions into an util file to reduce code duplication.